yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,683 @@
+
+E.g. if an identical transaction is submitted at time `t` and time `t+1`, the costs to the sequencer might be higher at `t+1` because the the gas price on L1 has increased, and thus verification costs are higher.
+
+We will need to develop a mechanism to ensure the user can specify a fee that is both reasonable and likely to be included in a block, and that the sequencer is compensated fairly for the costs they incur.
+
+Open discussions are being had around using a [Shapley cost sharing mechanism here](https://en.wikipedia.org/wiki/Shapley_value).
+
+In the immediate term, we will risk bad UX: the user will need to look at the `inclusion_fee` from previously included transactions to get a sense of what is reasonable, and thus likely overpay.
+:::
+
+# Fees per gas
+
+The fees per unit gas are set by the protocol, and are not part of the user's transaction.
+
+The fees are:
+
+- `fee_per_da_gas`: 1 `$AZT`
+- `fee_per_l2_gas`: 1 `$AZT`
+- `fee_per_l1_gas`: 100 `$AZT`
+
+:::note
+Current thinking is that these fees will eventually be dynamic, and set by the protocol based on demand, similar to eip-1559 style mechanics.
+
+E.g. [op stack directly scales off ethereum base fees for DA costs](https://docs.optimism.io/stack/transactions/fees#l1-data-fee)
+:::
+
+# Transaction fee
+
+A transaction fee is the maximum amount of `$AZT` that a user can be charged for a transaction.
+
+```
+transaction_fee = fee_per_da_gas * da_gas_used
+                    + fee_per_l2_gas * l2_gas_used 
+                    + fee_per_l1_gas * l1_gas_used
+                    + inclusion_fee
+```
+
+:::warning Fees are not enforced 
+It is not enforced by the protocol that the `transaction_fee` is actually paid. It is up to the sequencer to enforce this.
+
+The **only way** the `transaction_fee` is paid is if the transaction includes a "teardown" function that calls `pay_fee` on the enshrined gas token. [More on this below](#validating-fee-payment).
+
+When a transaction is published, the `transaction_fee` is included but is only non-zero if the fee was paid.
+:::
+
+### `l1_gas_used`
+
+`l1_gas_used` is the amount of gas consumed by the transaction while sending messages to L1.
+
+Each message sent to L1 consumes 1 unit of L1 gas.
+
+
+### `l2_gas_used`
+
+`l2_gas_used` is the amount of gas consumed by the transaction while executing the public VM.
+
+It is comprised of the sum of:
+- the sum of all gas consumed by the public VM viz opcode-based metering
+- the number of enqueued public function calls, multiplied by a fixed amount of gas per call
+
+The second component compensates the sequencer for the cost of proving the public VM circuits.
+
+:::warning
+Add link to l2 opcode metering
+:::
+
+
+### `da_gas_used`
+
+`da_gas_used` is the amount of DA gas consumed by the transaction.
+
+```
+da_gas_used = FIXED_DA_GAS +
+                note_hash_gas +
+                nullifier_gas +
+                l2_to_l1_message_gas +
+                public_data_writes_gas +
```

Comment:
What about the costs imposed on all other nodes (that aren't the sequencer creating this block)? I.e. the costs to pay the network to update their nodes' world state dbs and storing this information? (Or, rather than paying everyone (which obviously doesn't happen), it's more paying to outcompete anyone else who wants their tx to be included by the scarce resources of the network). This kind of cost feels very similar to why we pay for ethereum txs.

Reply:
Should the cost of a new public data write cost more than overwriting an existing public data leaf?

Reply:
Great points about adding data to the trees and their varying costs. I think we could add scalars to the computations below to compensate differently based on the work required for different insertions e.g.:
```
note_hash_gas = da_gas_per_field * (number of notes) * a
nullifier_gas = da_gas_per_field * (number of nullifiers) * b
```

I think doing something like that is predicated on being quite sure that the relative costs of insertions will be what we presently think they are